### PR TITLE
Fix conditional type constraint exploration + signature relations

### LINKED ISSUE #47142: Stack overflow in TS 4.5.3 and nightly
> I'm running into a similar issue. When I run `tsc` on our `node` project, I get `RangeError: Maximum call stack size exceeded`.
> 
> Everything was working fine - this just happens when I attempt to upgrade to typescript `4.5.3`. It works fine in typescript `4.2.3`.
> 
> FWIW, recursive types should definitely be allowed. They're a standard part of any programming language (more generally, Context Free Grammars) and it's very common to do this (e.g. with datastructures like trees, or compilers, etc.).
> 
> e.g. it's 100% valid to have:
> ```
> type Tree<T> = { data: T; children: Tree<T>[] } 
> ```
> 
> Unfortunately the stacktrace I get doesn't give me any indication of which type is causing the issue:
> ```
> /Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:94444
>                 throw e;
>                 ^
> 
> RangeError: Maximum call stack size exceeded
>     at getRelationKey (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:54317:32)
>     at recursiveTypeRelatedTo (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:52985:26)
>     at isRelatedTo (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:52645:30)
>     at typeRelatedToSomeType (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:52869:35)
>     at structuredTypeRelatedToWorker (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:53099:32)
>     at structuredTypeRelatedTo (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:53084:30)
>     at isRelatedTo (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:52642:30)
>     at checkTypeRelatedTo (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:52314:26)
>     at isTypeRelatedTo (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:52276:24)
>     at isTypeAssignableTo (/Users/cbattycapps/{REDACTED}/node_modules/typescript/lib/tsc.js:51558:20)
> ```

_Originally posted by @charlescapps in https://github.com/microsoft/TypeScript/issues/45576#issuecomment-991376337_

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #47142.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
Adding a regression test isn't simple as the repro references multiple NPM packages.

**Comment by ahejlsberg** [comment]
To be clear, this PR fixes the repro here https://github.com/microsoft/TypeScript/issues/47142#issuecomment-996796794.

**Review by weswigham — APPROVED** [review]
Individualally, these two changes look fine and make sense on their face - the first skip distributive constraints for circularly constrained types, which is good, and the second expands simplified signature matching to instances of the same type reference rather than just instances if the same symbol... which it's weird that the second isn't fully encompassed by the first, which is why an example and repro would be nice, but 🤷‍♂️

**Comment by ahejlsberg** [comment]
> it's weird that the second isn't fully encompassed by the first

The change to not explore circular constraint is what fixes the stack overflow. Definitely a good change.

The signature relation change keeps us from cross-comparing incompatible signatures. Without it, as we're computing the variance for `UpsertGraphMethod` (which has multiple signatures), we relate a `PartialModelGraph<M>` to a `PartialModelGraph<M>[]` (all are types from the [`Objection.js`](https://github.com/vincit/objection.js)  NPM package). That then sets off a cascade of deeper and deeper nested complex combinations of conditional types, index access types, and intersections. Deep enough, I guess, that `isDeeplyNestedType` never sees three cycles and we eventually hit the excessive stack depth limiter.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -19920,7 +19920,8 @@ namespace ts {
                 const incompatibleReporter = kind === SignatureKind.Construct ? reportIncompatibleConstructSignatureReturn : reportIncompatibleCallSignatureReturn;
                 const sourceObjectFlags = getObjectFlags(source);
                 const targetObjectFlags = getObjectFlags(target);
-                if (sourceObjectFlags & ObjectFlags.Instantiated && targetObjectFlags & ObjectFlags.Instantiated && source.symbol === target.symbol) {
+                if (sourceObjectFlags & ObjectFlags.Instantiated && targetObjectFlags & ObjectFlags.Instantiated && source.symbol === target.symbol ||
+                    sourceObjectFlags & ObjectFlags.Reference && targetObjectFlags & ObjectFlags.Reference && (source as TypeReference).target === (target as TypeReference).target) {
                     // We have instantiations of the same anonymous type (which typically will be the type of a
                     // method). Simply do a pairwise comparison of the signatures in the two signature lists instead
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:19926]: Is this comment accurate (i.e. "same anonymous type")?

**Comment by sandersn** [comment]
On DT, breaks knockout.mapping with

```
Type 'KnockoutObservable<false> | KnockoutObservable<true>' is not assignable to type 'KnockoutObservable<boolean>'
```

Is this expected? I'll add more detail after investigating.

**Comment by sandersn** [comment]
Here's a small repro. I'll file it as a separate bug so we can decide whether it's worthwhile to fix:

```ts
interface Kow<T> {
    (value: T): void;
}
declare var src: Kow<true> | Kow<false>;
var tgt: Kow<boolean> = src
```

**Comment by ahejlsberg** [comment]
Marking this as a breaking change since the error in #47371 is the intended behavior.

**Pull request merged**